src/System.Linq.Expressions/src/System/Linq/Expressions/MethodCallExpression.cs

```diff
@@ -879,6 +1042,34 @@ public static MethodCallExpression Call(Type type, string methodName, Type[] typ
         ///<paramref name="instance" />.Type is not assignable to the declaring type of the method represented by <paramref name="method" />.-or-The number of elements in <paramref name="arguments" /> does not equal the number of parameters for the method represented by <paramref name="method" />.-or-One or more of the elements of <paramref name="arguments" /> is not assignable to the corresponding parameter for the method represented by <paramref name="method" />.</exception>
         public static MethodCallExpression Call(Expression instance, MethodInfo method, IEnumerable<Expression> arguments)
         {
+            IReadOnlyList<Expression> argumentList = arguments as IReadOnlyList<Expression>;
```

Comment:
Note this differs a tiny bit from the approach taken in BlockExpression, where the optimization is only in place for the Expression[] overload. That has limited mileage, because various other paths (such as Rewrite and Update) call into the IEnumerable<Expression> overload. Also note I'm using IReadOnlyList<T> which is covariant in T, so even passing in a compatible collection of a more derived type will pass the type test.

The one little change in behavior here is the virtual calls to the Count property and indexer if the test passes, therefore relying on the list to be well-behaved in terms of the interface contract. To be fair, existing code behind ToReadOnly does a similar check based on ICollection<T> (less applicable given it's not covariant in T, maybe worth looking into separately) so such assumptions of well-behaved interface implementations exist elsewhere.